core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java

```diff
@@ -457,6 +487,39 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
         return null;
     }
 
+    private void interruptRunawaySplits()
+    {
+        for (RunningSplitInfo splitInfo : runningSplitInfos) {
+            Duration duration = Duration.succinctNanos(ticker.read() - splitInfo.getStartTime());
+
+            if (duration.compareTo(interruptRunawaySplitsTimeout) > 0) {
+                List<StackTraceElement> stackTraceElements = asList(splitInfo.getThread().getStackTrace());
+
+                if (!splitInfo.printed) {
+                    splitInfo.setPrinted();
+
+                    String splitThreadStack = toStackString(stackTraceElements);
+                    String splitThreadId = splitInfo.getThreadId();
+                    log.warn(String.format("Split %s has been long running %n%s", splitInfo.getSplitInfo(), splitThreadStack));
+                }
+
+                if (interruptibleSplitPredicate.test(stackTraceElements)) {
+                    log.warn("Interrupting runaway split matching the predicate " + splitInfo.getSplitInfo());
+                    splitInfo.getThread().interrupt();
```

Comment:
is this racy?

Reply:
That is a good point. 
We need coordination between a code that issues an interruption and runner threads. So we know that we are only interrupting the runner thread if it is still executing the query we observed in the stack trace (and not after the thread picked up split from another query).

Also does interrupting thread which executes within Joni library actually makes the call from the thread return?

Reply:
> Also does interrupting thread which executes within Joni library actually makes the call from the thread return?

currently seems that it catches the interruptedexception [silently](https://github.com/airlift/joni/blob/master/src/org/joni/Matcher.java#L96) with an error value that does not seem handled. would need something like https://github.com/prestodb/presto/pull/16109 . but this may not be guaranteed based on this comment https://github.com/prestodb/presto/pull/16109#pullrequestreview-661488484